dcicutils/contribution_utils.py

```diff
@@ -0,0 +1,532 @@
+            # This will set .loaded_contributor_data and other values from CONTRIBUTORS.json
+            self.load_contributors_from_json_file_cache(existing_contributor_data_file)
+
+        if exclude_fork and not self.excluded_contributions:
+            self.excluded_contributions = Contributions(repo=exclude_fork)
+        checkpoint1 = self.checkpoint_state()
+        self.reconcile_contributors_with_github_log()
+        checkpoint2 = self.checkpoint_state()
+
+        def list_to_dict_normalizer(*, label, item):
+            ignored(label)
+            if isinstance(item, list):
+                return {elem: elem for elem in item}
+            else:
+                return item
+
+        if existing_contributor_data_file:
+            diff_manager = DiffManager(label="contributors")
+            contributors1 = checkpoint1['contributors_by_name']
+            contributors2 = checkpoint2['contributors_by_name']
+            diffs = diff_manager.diffs(contributors1, contributors2, normalizer=list_to_dict_normalizer)
+            added = diffs.get('added')
+            changed = diffs.get('changed')
+            removed = diffs.get('removed')
+            self.cache_discrepancies = cache_discrepancies = {}
+            if added:
+                cache_discrepancies['to_add'] = added
+            if changed:
+                cache_discrepancies['to_change'] = changed
+            if removed:
+                cache_discrepancies['to_remove'] = removed
+
+    def load_contributors_from_json_file_cache(self, filename):
+        self.loaded_contributor_data = data = self.get_contributors_json_from_file_cache(filename)
+        self.load_from_dict(data)
+
+        if DEBUG_CONTRIBUTIONS:  # pragma: no cover - debugging only
+            PRINT("After load_contributors_from_json_file_cache...")
+            PRINT(f"{n_of(self.pre_fork_contributors_by_name, 'pre-fork contributor by name')}")
+            PRINT(f"{n_of(self.pre_fork_contributors_by_email, 'pre-fork contributor by email')}")
+            PRINT(f"{n_of(self.contributors_by_name, 'contributor by name')}")
+            PRINT(f"{n_of(self.contributors_by_email, 'contributor by email')}")
+
+    def reconcile_contributors_with_github_log(self):
+        """
+        Rummages the GitHub log entries for contributors we don't know about.
+        That data is merged against our existing structures.
+        """
+        if DEBUG_CONTRIBUTIONS:  # pragma: no cover - debugging only
+            PRINT("Reconciling with git log.")
+        excluded_fork_contributors_by_email = (self.excluded_contributions.contributors_by_email
+                                               if self.excluded_contributions
+                                               else {})
+        # excluded_fork_contributors_by_name = (self.excluded_contributions.contributors_by_name
+        #                                       if self.excluded_contributions
+        #                                       else {})
```

Comment:
It was something I was experimenting with. I only need email but I should take in the _by_name data and convert it to _by_email because the _by_name data is more compact and more intelligible. I'll work on that later.